### PR TITLE
Volume cloning feature

### PR DESCRIPTION
**Pull request by amacaskill**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
This PR implements volume cloning. 
**Which issue(s) this PR fixes**:

Fixes #https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/161

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Enables volume cloning.
```

### CODE REVIEW

File: pkg/gce-pd-csi-driver/controller_test.go [hunk 1]
```diff
@@ -900,6 +900,114 @@ func TestCreateVolumeWithVolumeSource(t *testing.T) {
 	}
 }
 
+func TestCreateVolumeWithVolumeSourceFromVolume(t *testing.T) {
+	testSourceVolumeName := "test-volume-source-name"
+	testVolumeSourceID := fmt.Sprintf("projects/%s/zones/%s/disks/%s", project, zone, testSourceVolumeName)
+	testVolumeSourceIDDifferentZone := fmt.Sprintf("projects/%s/zones/%s/disks/%s", project, "different-zone", testSourceVolumeName)
+	// Define test cases
+	testCases := []struct {
```
[the remark this answers is not in the document]
Comment (amacaskill) [reply pkg/gce-pd-csi-driver/controller_test.go:908]: what do you mean? The create volume request with either have 

```VolumeContentSource: &csi.VolumeContentSource{
				Type: &csi.VolumeContentSource_Volume{
					Volume: &csi.VolumeContentSource_VolumeSource{
						VolumeId: tc.sourceVolumeID,
					},
				},
```
or 

```VolumeContentSource: &csi.VolumeContentSource{
				Type: &csi.VolumeContentSource_Snapshot{
					Snapshot: &csi.VolumeContentSource_SnapshotSource{
						SnapshotId: snapshotID,
					},
				},
```
then in the CreateVolume function, depending on whatever is in the `content := req.GetVolumeContentSource()` we either change snapshotID to have a snapshotID or volumeContentSourceVolumeID to have a volumeID. Did you mean something like this? 
```
                                        Volume: &csi.VolumeContentSource_VolumeSource{
						  VolumeId: testSnapshotID,
					},
```

[the remark this answers is not in the document]
Comment (mattcary) [reply pkg/gce-pd-csi-driver/controller_test.go:908]: I mean make a test with an invalid source that specifies both the source volume & snapshot ids.

[the remark this answers is not in the document]
Comment (mattcary) [reply pkg/gce-pd-csi-driver/controller_test.go:908]: Discussed offline, I was confused, this case is not possible by the syntax of VolumeContentSource.

File: pkg/gce-pd-csi-driver/controller_test.go [hunk 2]
```diff
@@ -900,6 +900,114 @@ func TestCreateVolumeWithVolumeSource(t *testing.T) {
...
+		{
+			name:           "fail with data source of volume type with invalid disk parameters",
+			project:        "test-project",
+			volKey:         meta.ZonalKey("my-disk", zone),
+			volumeOnCloud:  true,
+			expErrCode:     codes.InvalidArgument,
+			sourceVolumeID: testVolumeSourceIDDifferentZone,
+			parameters: map[string]string{
+				common.ParameterKeyType: "different-type",
+			},
```
Comment (mattcary) [inline pkg/gce-pd-csi-driver/controller_test.go:949]: Add tests with regionalKey?

File: pkg/gce-pd-csi-driver/controller.go [hunk 3]
```diff
@@ -204,24 +204,60 @@ func (gceCS *GCEControllerServer) CreateVolume(ctx context.Context, req *csi.Cre
...
+				}
+			}
+
+			//	Verify the zone, region, and disk type of the clone must be the same as that of the source disk.
+			if err := gce.ValidateDiskParameters(diskFromSourceVolume, params); err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, `CreateVolume source volume parameters do not match CreateVolumeRequest Parameters: %v`, err)
+			}
+
+			// Verify the source disk is ready.
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (amacaskill) [reply pkg/gce-pd-csi-driver/controller.go]: Oh good catch no that if check doesn't do anything I removed it.

File: deploy/kubernetes/images/prow-gke-release-staging-rc-master/image.yaml [hunk 4]
```diff
@@ -52,6 +52,5 @@ metadata:
   name: imagetag-csi-gce-driver-prow-rc
 imageTag:
   name: k8s.gcr.io/cloud-provider-gcp/gcp-compute-persistent-disk-csi-driver
-  newName: gcr.io/k8s-staging-cloud-provider-gcp/gcp-compute-persistent-disk-csi-driver
-  newTag: "v1.3.4-rc1"
+  newTag: "v1.3.4"
```
[the remark this answers is not in the document]
Comment (amacaskill) [reply deploy/kubernetes/images/prow-gke-release-staging-rc-master/image.yaml]: Something happened the first time I tried squashing my commits, but that is reverted now.

File: test/k8s-integration/config/test-config-template.in [hunk 5]
```diff
@@ -4,6 +4,11 @@ StorageClass:
 SnapshotClass:
   FromFile: {{ .SnapshotClassFile }}
 {{end}}
+{{if .Timeouts}}
```
Comment (mattcary) [inline test/k8s-integration/config/test-config-template.in:7]: These timeouts are no longer needed since we're just skipping the tests?

Comment (amacaskill) [reply test/k8s-integration/config/test-config-template.in]: removed

File: test/k8s-integration/main.go [hunk 6]
```diff
@@ -543,6 +543,9 @@ func generateGCETestSkip(testParams *testParameters) string {
 	if testParams.platform == "windows" {
 		skipString = skipString + "|\\[LinuxOnly\\]"
 	}
+	// Volume cloning has timeouts due to GCE disk cloning rate limits, and operation serialization
+	// race conditions where the cloning begins while the source disk is still being created.
+	skipString = skipString + "|provisioning\\sshould\\sprovision\\sstorage\\swith\\spvc\\sdata\\ssource[^|]*"
```
Comment (msau42) [inline test/k8s-integration/main.go:548]: Is this a problem with the OSS test that can be fixed?

Comment (msau42) [reply test/k8s-integration/main.go:548]: Does this mean we don't have an e2e test exercising this?

Comment (amacaskill) [reply test/k8s-integration/main.go:548]: There is a race condition in the kubernetes e2e test that is causing timeouts. This PR fixes the problem making the tests pass: https://github.com/kubernetes/kubernetes/pull/106322, but it has not been merged yet. So we are skipping those tests until that PR gets merged.

Comment (mattcary) [reply test/k8s-integration/main.go:548]: We're going to have to write a guitar test to cover the gap. We're talking with Jan on fixing upstream test coverage but I don't want to block on it.

**Pull request merged**